ethcore/src/verification/queue/mod.rs

```diff
@@ -505,6 +506,16 @@ impl<K: Kind> VerificationQueue<K> {
 				match err {
 					// Don't mark future blocks as bad.
 					Error(ErrorKind::Block(BlockError::TemporarilyInvalid(_)), _) => {},
+					// If the transaction root or uncles hash is invalid, it doesn't necessarily mean
+					// that the header is invalid. We might have just received a malformed block body,
+					// so we shouldn't put the header hash to `bad`.
+					//
+					// We still put the entire `Item` hash to bad, so that we can early reject
+					// the items that are malformed.
+					Error(ErrorKind::Block(BlockError::InvalidTransactionsRoot(_)), _) |
+					Error(ErrorKind::Block(BlockError::InvalidUnclesHash(_)), _) => {
+						self.verification.bad.lock().insert(raw_hash);
+					},
 					_ => {
 						self.verification.bad.lock().insert(hash);
```

Comment:
@tomusdrw I'm thinking maybe it's better to always track bad blocks by `raw_hash`?

Reply:
I think it may require either putting both hashes in or changing some sync code as we expose `status(H256) -> Status` method, so that other components can check by-hash if particular header was already marked as bad.
I didn't want to change too much in that PR, but I think it may indeed make sense long term.